### ISSUE TITLE
Development Version

### ISSUE THREAD

**Issue by Thrameos** [originell/jpype#329]
I am once again closing on a version 0.7.  The version addresses the following core issues.
- Unicode handling between python and java.
- Class organization
- Error reporting 
- Removal of dead code
- Simplification of code paths.
- Requisite tasks for switching to Java classloader extensions.
- Removal of Capsule logic.

As part of the work I have identified a few problematic features that likely should be dealt with.  Unfortunately they are all API breakers, thus it would be a good idea to get input from users.

1.  Exceptions - Currently we don't throw java exceptions a python exceptions but rather a wrapper contained within.  This leads to confusing and hard to follow exception handling.  I can find no reason for this construct in Python 2.7 nor Python 3.6.  I am guessing it is from a much earlier point in time.  Removing this extra layer would make handling of exceptions easier.

2. Bean pattern - Currently a Class customizer is going through and added all "get/set" patterns as properties. Unfortunately, this has been a huge source of bugs in my development team.  The properties override the actual field and thus make debugging difficult as you think you are accessing the field but instead calling methods.  This is also a violation of Python's motto "There should only be one way to do things."   Removing this misfeature would potentially break code.  About the only way to have this feature and be able not have its problematic default behavior would be to have a "import jpype.beans" that holds this problematic customizer to support older code. 

3. String conversion - This one has also been a source of bugs and incompatiblity.  JPype converts strings on return to python rather than giving a proper java.lang.String.  This removes the expected string methods and make it so that even when we are manipulation java strings purely by passing back and forth between java we are paying conversion costs.  There is an option to disable this behavior, but it is global.  Modules written using jpype that assume one behavior fight with modules written the other way.  This has been a huge headache.  There is not many good ways to support both modes and not create problems.  And the default always convert is the worst in terms of performance. The existing code gives a half converted object JStringWrapper, but the refactored code gives a full java.lang.String.   The impact depends on how much manipulation the string gets manipulated after it is returned.  It would be best to remove the autoconversion misfeature.  At minimum the default needs to switch to the more consistent keep java objects as handles and the old autoconvert needs to become an option with a deprecated message.

Although none of these are fatal, they all cause difficulties in use of jpype.  I generally prefer keeping the API the same, but if a misfeature gets added we should deprecate it and provide a path to a cleaner design. 

@marscher at some point we should put a devel branch in the master repo so that I can pull request to that rather than the main.  Currently I have to pull request to master to get the CI checks.  There is going to be a period in which the development version is functional, but still not leak free nor stable enough for production.

**Comment by marscher** [comment]
Thanks for clarifying all the changes and problems we have right now. I
think you are really pushing the project into the right direction!

I can add a devel branch on your behalf or I just give the right to do
it (if you wish).

**Comment by Thrameos** [comment]
Better to have you make the devel branch.  I am not well versed in getting CI linked in github.  I am not sure I want direct access as I sometimes go off on bad tangents.

**Comment by marscher** [comment]
there you go, it is named 'devel'

**Comment by Thrameos** [comment]
So do you have any opinions on the misfeatures?   I was thinking one way to support it would be to make a `import jpype.deprecated`  as a package option which would switch everything to the previous defaults.  That way someone would just need to add that import to use old code with the new version.  Otherwise I am going to have haul earlier bad decision forward forever.

**Comment by Thrameos** [comment]
Submitted as #334
